xdsl/dialects/csl.py

```diff
@@ -1511,6 +1565,18 @@ def verify_(self) -> None:
                 "Input and output type must be of different signedness"
             )
 
+    def __init__(self, op: SSAValue | Operation):
+        typ = op.results[0].type if isinstance(op, Operation) else op.type
+        assert isinstance(typ, IntegerType)
+        res_signedness = (
+            Signedness.SIGNLESS
+            if typ.signedness.data == Signedness.UNSIGNED
+            else Signedness.UNSIGNED
+        )
+        super().__init__(
+            operands=[op], result_types=[IntegerType(typ.width, res_signedness)]
+        )
```

Comment:
I think I'd prefer a much simpler interface that is in-line with other cast operations:

```suggestion
    def __init__(self, op: SSAValue | Operation, result_type: IntegerType):
        super().__init__(
            operands=[op], result_types=[result_type]
        )
```

This way the caller can decide what he wants instead of the init smartly guessing.

Reply:
If you prefer a simple interface, I suggest it's simplest to do what the op promises to do, and just flip the signedness.

Reply:
But you may want to cast from `i32` to `si32`, which this constructor does not allow.

Reply:
I understood we didn't want to have `si32`, but I added an optional arg to explicitly state the result type, and if it's `None` we derive it.